### ISSUE TITLE
S3 Native State Locking -- "Content-MD5 OR x-amz-checksum- HTTP header is required for Put Object requests with Object Lock parameters"

### ISSUE THREAD

**Issue by BroMattMiller** [hashicorp/terraform#36113]
### Terraform Version

```shell
Terraform v1.10.0
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v5.78.0
```

### Terraform Configuration Files

```terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
    }
  }

  backend "s3" {
    bucket         = "state.test.biblestudy.tools"
    key            = "terraform.tfstate"
    region         = "us-east-1"
    use_lockfile   = true
  }
}

```

### Debug Output

https://gist.github.com/BroMattMiller/54b3c5ab4ac2f052157daa66536d392e

### Expected Behavior

After moving to native S3 locking the apply should have succeeded.

### Actual Behavior

The apply failed.

### Steps to Reproduce

```
terraform init
terraform apply
```

**Comment by bschaatsbergen** [comment]
Hey @BroMattMiller,

Thank you for reporting this! The S3 backend is managed by the AWS Provider team at HashiCorp, and this issue has been added to their triage queue. Thanks again!

**Comment by BroMattMiller** [comment]
Yes, I have enabled Object Lock, per AWS Security Hub control S3.15, "S3 general purpose buckets should have Object Lock enabled." The retention mode is Governance and the retention date is 30 days. The credentials used be Terraform do have the necessary privileges to override this.

**Comment by bschaatsbergen** [comment]
When S3 Object Lock is enabled on a bucket with a retention period, S3 requires the Content-MD5 or x-amz-sdk-checksum-algorithm header for object uploads (via PutObject) to ensure data integrity during the upload process.

Terraform’s state writes to the S3 bucket relied on an [uploader](https://github.com/aws/aws-sdk-go-v2/blob/8939ed049c8b3490f9a3dbd4e4d3d56cf22a27f2/feature/s3/manager/upload.go#L193-L200) from [aws-sdk-go-v2](https://github.com/aws/aws-sdk-go-v2), which appends these required headers by automatically computing the checksum. However, the lock file implementation did not use this uploader, resulting in missing headers for PutObject requests and conflicts with Object Lock requirements.

**Comment by bschaatsbergen** [comment]
Hey @BroMattMiller,

This should now be fixed in Terraform 1.10.1

**Comment by BroMattMiller** [comment]
Yep, I just now tested it, and we’re good. 

Thanks.
